CARDS-1886: Replace Torch import with Clarity import

sdumitriu opened the pull request:
Also includes CARDS-2208: Clarity import: NOT IN rules for configured mappers and filters don't work as expected

To test proms:

1. build with `mvn clean install -Pdocker`
2. in `compose-cluster/mssql` run `python3 generate_test_proms_sql.py -n 100 proms_sample.sql`
3. in `compose-cluster` run `python3 generate_compose_yaml.py --mssql --cards_project cards4proms --oak_filesystem --dev_docker_image --smtps --smtps_test_container --smtps_test_mail_path ~/cardsmail --composum --adminer --clarity`
4. in `compose-cluster` start cards with `docker-compose up`
5. at `http://localhost:1435/?mssql=mssql&username=sa&db=master&ns=path&import=` import the generated `proms_sample.sql` file
6. open `http://localhost:8080/Subjects.importClarity`
7. check that the patients have been imported correctly
8. stop docker-compose and run `docker-compose rm` and `./cleanup.sh`
9. repeat steps 3-5
10. at `http://localhost:8080/system/console/configMgr` search for `NightlyClarityImport` and change the configuration to have the import schedule at about 2 minutes from now (the format is `seconds minute hour ...`, so change it to something like `0 45 14 * * ? *` to run at `14:45`)
11. wait for 2 minutes, then check that the forms have been imported

Regression testing for prems:

1. build with `mvn clean install -Pdocker`
2. in `compose-cluster/mssql` run `python3 ./generate_test_YE_sql.py -n 100 --basedate 2023-04-24 --time_spread_seconds 0 prems_sample.sql` (use 7 days ago as the date)
3. in `compose-cluster` run `python3 generate_compose_yaml.py --mssql --cards_project cards4prems --oak_filesystem --dev_docker_image --smtps --smtps_test_container --smtps_test_mail_path ~/cardsmail --composum --adminer --clarity`
4. in `compose-cluster` start cards with `docker-compose up`
5. at `http://localhost:1435/?mssql=mssql&username=sa&db=master&ns=path&import=` import the generated `prems_sample.sql` file
6. open `http://localhost:8080/Subjects.importClarity?dayToQuery=-7`
7. check that the patients have been imported correctly
8. stop docker-compose and run `docker-compose rm` and `./cleanup.sh`
9. repeat steps 3-5
10. at `http://localhost:8080/system/console/configMgr` search for `NightlyClarityImport` and change the configuration to have the import schedule at about 2 minutes from now (the format is `seconds minute hour ...`, so change it to something like `0 45 14 * * ? *` to run at `14:45`)
11. wait for 2 minutes, then check that the forms have been imported


To do:
- [x] Rename the clinic ID to `TG-PMCC CARDIAC CLINICS DIAGNOSTIC`
- [x] Copy the clinic name to the Visit Information/Location answer
- [x] Special handling for Canceled events: if previously imported, update the existing Visit Information/Status answer 
- [x] Discard visits without DOB or both `MRN` and `OHIP_NBR`
- [x] Add a processor that automatically maps between `ENCOUNTER_CLINIC` and Visit Information/Clinic

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Move the Torch Import code into a separate module

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Remove configurations for the torch importer from proms and prems

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Make the visit identifier available under a common name instead of relying on the specific Clarity view

sdumitriu committed:
CARDS-2208: Clarity import: NOT IN rules for configured mappers and filters don't work as expected

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Make generic clarity processors easy to enable/disable

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Use the common subject identifier "column" instead of always configuring it to the Clarity source

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Make the EmailConsentFilter more customizable

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Move generic filters and mappers into the common clarity-integration module

sdumitriu committed:
[misc] Rename FilterEmailConsent to EmailConsentFilter to be consistent with all the other filter class names

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Add the Clarity column mapping

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Remove the MOOD clinic for now

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Enable the Clarity feature when the proms project is selected

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Basic Clarity data filters and mappers

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Add script for generating sample proms clarity data

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Add support for importing *future* events from Clarity, not just past events

sdumitriu committed:
[misc] Refactoring to remove useless assignment

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Allow special characters in the Clarity table name

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Add support for Proms on Clarity in the docker compose generation script

sdumitriu commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Survey/Mood.xml:
@marta- You want to keep this?

sdumitriu commented on proms-resources/feature/src/main/features/feature.json:
```diff
@@ -85,6 +82,52 @@
+    "io.uhndata.cards.clarity.importer.internal.EmailConsentFilter": {
+      "enable": true,
+      "emailColumn": "EMAIL",
+      "emailConsentColumn": ""
```
Email consent validation is disabled for now, should I enable it?

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Email consent is implied when MyChart is activated

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Change the Cardio clinic name to TG-PMCC CARDIAC CLINICS DIAGNOSTIC

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Survey/ClinicMapping.xml:
```diff
@@ -26,11 +26,11 @@
         <type>String</type>
     </property>
     <node>
-        <name>853703519</name>
```
Did you check that this doesn't break the clinic admin ui?

sdumitriu replied:
Yes, it does not.

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Copy the assigned clinic name to the Visit Information/Location answer

sashaandjic commented:
proms mode: step 11 - no import

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Discard visits without DoB or both MRN and OHIP_NBR

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Add a processor that automatically maps between ENCOUNTER_CLINIC and Visit Information/Clinic

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Special handling for Canceled events: if previously imported, update the existing Visit Information/Status answer

sashaandjic commented:
testing in progress:
**proms** mode: 
- [x] `steps 1-7`: all good ( 89 out of 100 visits imported, 11 visits cancelled); 623 forms generated
- [x] `steps 8-11`: all good - 89 visits; 623 forms
-----

sashaandjic reviewed (approved):
tested - all good!

sorry for the long comment, there was a lot to test in this PR

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/DiscardCanceledEvents.java:
```diff
@@ -0,0 +1,63 @@
+    public Map<String, String> processEntry(Map<String, String> input)
+    {
+        final String status = input.get("ENCOUNTER_STATUS");
+        if ("Canceled".equals(status) && this.trrp.getThreadResourceResolver().getResource(
```
`equalsIgnoreCase`

marta- commented on proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/MychartEmailConsentMapper.java:
```diff
@@ -17,43 +17,40 @@
  */
 
-package io.uhndata.cards.prems.internal.importer;
+package io.uhndata.cards.proms.internal.importer;
```
Moved or copied?

marta- commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/apps/cards/clarityMapping.xml:
```diff
@@ -0,0 +1,334 @@
+                <type>String</type>
+              </property>
+            </node>
+            <node>
+              <name>00000006</name>
+              <primaryNodeType>cards:ClarityQuestionMapping</primaryNodeType>
+              <property>
+                <name>column</name>
+                <value>ENCOUNTER_CLINIC</value>
+                <type>String</type>
+              </property>
+              <property>
+                <name>question</name>
+                <value></value>
+                <type>String</type>
+              </property>
+            </node>
+            <node>
```
Can this be merged into the clinic mapper?

marta- reviewed (commented):
To do:
* add mapper for EMAIL_CONSENT transforming string into "Yes" and null is No
* make sure cancelled events are correctly excluded on the clinic dashboard and by the email task

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Add a conversion between the custom UHN_EXTERNAL_EMAIL and Yes value for the EMAIL_CONSENT column

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Change custom Epic event statuses into FHIR standard values

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Merge ENCOUNTER_CLINIC and CLINIC into only one mapping column

sdumitriu committed:
CARDS-1886: Replace Torch import with Clarity import
Update the sample sql generator to match production data

sdumitriu commented:
> To do:
> 
>     * add mapper for EMAIL_CONSENT transforming string into "Yes" and null is No
> 
>     * make sure cancelled events are correctly excluded on the clinic dashboard and by the email task

All done.

sdumitriu commented:
Tested on production data, all is well.

sashaandjic reviewed (approved):
tested
